src/big/big_decimal.cr

```diff
@@ -0,0 +1,249 @@
+  include Comparable(String)
+
+  private property value : BigInt
+  private property scale : UInt64
+  getter value, scale
+
+  # Sets maximum iterations used in division operation. This implicitly
+  # defines the maximum precision of divisions in cases where the
+  # division is not exact.
+  def self.set_max_div_iterations(new_max_div : Int)
+    @@max_div_iterations = new_max_div.to_u64
+  end
+
+  # Creates a new `BigDecimal` from a `String`.
+  #
+  # Allows only valid number strings with an optional negative sign.
+  def initialize(str : String)
+    # disallow every non-valid string
+    if str !~ /^-?[0-9]+(\.[0-9]+)?$/
+      raise InvalidBigDecimalException.new("Invalid BigDecimal: #{str}")
+    end
+
+    if str.includes?('.')
+      v1, v2 = str.split('.')
+      @value = "#{v1}#{v2}".to_big_i
+      @scale = v2.size.to_u64
+    else
+      @value = str.to_big_i
+      @scale = 0_u64
+    end
+  end
+
+  # from `Int`
+  def initialize(num : Int)
+    initialize num.to_s
+  end
+
+  # from `Float` is not supported due to precision loss risks. This call fails at compile time.
+  def initialize(num : Float)
+    {% raise "Initializing from Float is risky due to loss of precision -- use Int, String or BigRational" %}
+    initialize num.to_s # to appease the compiler
+  end
+
+  # from `BigInt`/`UInt64` which matches the internal representation
+  def initialize(value : BigInt, scale : UInt64)
+    @value = value
+    @scale = scale
+  end
+
+  # from `Int`/`Int`, converting to internal representation
+  def initialize(value : Int, scale : Int)
+    initialize(value.to_big_i, scale.to_u64)
+  end
+
+  def initialize(value : BigInt)
+    initialize(value, 0u64)
+  end
+
+  # zero
+  def initialize
+    initialize(0)
+  end
+
+  def +(other : BigDecimal) : BigDecimal
+    if @scale > other.scale
+      scaled = other.scale_to(self)
+      BigDecimal.new(@value + scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value + other.value, other.scale)
+    else
+      BigDecimal.new(@value + other.value, @scale)
+    end
+  end
+
+  def -(other : BigDecimal) : BigDecimal
+    if @scale > other.scale
+      scaled = other.scale_to(self)
+      BigDecimal.new(@value - scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value - other.value, other.scale)
+    else
+      BigDecimal.new(@value - other.value, @scale)
+    end
+  end
+
+  def *(other : BigDecimal) : BigDecimal
+    BigDecimal.new(@value * other.value, @scale + other.scale)
+  end
+
+  def /(other : BigDecimal) : BigDecimal
+    check_division_by_zero other
+
+    scale = @scale - other.scale
+    n, d = @value, other.@value
+
+    quotient, remainder = n.tdiv(d), n.remainder(d)
+    return BigDecimal.new(quotient, scale) if remainder == ZERO
+
+    remainder = remainder * TEN
+
+    i = 0
+    while remainder != ZERO && i < @@max_div_iterations
```

Comment:
Shouldn't you raise if you reach max_div_iterations, not give the wrong result.

Reply:
It's not very exceptional, since  simple calculations like 1/3 hit this.

Reply:
Oh never mind then. I didn't read the code properly.